Algorithm.Framework/Portfolio/MeanVarianceOptimizationPortfolioConstruction.py

```diff
@@ -0,0 +1,136 @@
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from clr import AddReference
+AddReference("System")
+AddReference("QuantConnect.Algorithm")
+AddReference("QuantConnect.Common")
+AddReference("QuantConnect.Indicators")
+
+from System import *
+from QuantConnect import *
+from QuantConnect.Util import PythonUtil
+from QuantConnect.Algorithm import *
+from QuantConnect.Algorithm.Framework import *
+from QuantConnect.Algorithm.Framework.Alphas import *
+from QuantConnect.Data.UniverseSelection import *
+from datetime import timedelta
+import numpy as np
+import pandas as pd
+from scipy.optimize import minimize
+
+### <summary>
+### Provides an implementation of Mean-Variance portfolio optimization based on modern portfolio theory.
+### The target percent holdings of each security is calculated by maximizing the Sharpe ratio. 
+### The interval of weights in optimization method can be changed based on the long-short algorithm.
+### The default model uses the last three months daily price to calculate the optimal weight 
+### with the weight range from -1 to 1  
+### </summary>
+class MeanVarianceOptimizationPortfolioConstructionModel:
+    def __init__(self, resolution=Resolution.Daily, interval=63, min_weight=-1, max_weight=1):
+        """ 
+        Initialize the model
+        Args:
+            resolution:The resolution of the history price 
+            interval(int): The time interval of history price to calculate the weight 
+            min_weight(float): The lower bounds on portfolio weights
+            max_weight(float): The upper bounds on portfolio weights
+        """
+        self.securities = [ ]
+        self.resolution = resolution
+        self.interval = interval
+        self.min_weight = min_weight
+        self.max_weight = max_weight
+
+    def CreateTargets(self, algorithm, insights):
+        """ 
+        Create portfolio targets from the specified insights
+        Args:
+            algorithm:The algorithm instance
+            insights: The insights to create portoflio targets from
+        Returns: 
+            An enumerable of portoflio targets to be sent to the execution model
+        """
+        if len(self.securities) == 0:
+            return [ ]
+        
+        symbols = [insight.Symbol for insight in insights]
+        # request the daily history price of last year
+        hist = algorithm.History(symbols, self.interval, self.resolution)
+        df_price = hist['close'].unstack(level=0)
+        # calculate the daily return 
+        daily_return = (df_price / df_price.shift(1)).dropna()
+        weights = PortfolioOptimization(daily_return, self.min_weight, self.max_weight).opt_portfolio() 
```

Comment:
Thinking out loud here...

My understanding of MPT is that it centers around the concept of *expected* returns. While it is common to use historical returns, we're actually generating predictions, some of which will have a magnitude which _is_ the expected return for the asset. I'm thinking it may make sense to use the magnitude from the insight if available as the expected return, and if not available, can fall back on historical returns as a proxy. I'm not sure how that would tie into what's here, but IMO it's worth consideration.

Reply:
Yes. In theory, MPT should have an expected return for each asset which is commonly the mean of historical return. Then the objectives are considered to minimize risk given a specific return. That's one of the problems MPT can solve. Another one is maximizing return for a given level of risk. Here I try to solve a numerical optimization problem to optimize both the return the variance (maximize the Sharpe ratio).